services/shell/copy.go

```diff
@@ -0,0 +1,112 @@
+package shell
+
+import (
+	"context"
+	"io/fs"
+
+	servicepb "go.viam.com/api/service/shell/v1"
+)
+
+// A File is a complete File to be sent/received.
+type File struct {
+	// RelativeName is the filesystem agnostic name.
+	// For example, ~/my/file would appear as my/file.
+	RelativeName string
+	// The underlying data for the File and it must have
+	// a non-nil fs.FileInfo.
+	Data fs.File
+}
+
+// A FileCopier receives files to copy into some implementation specific
+// location. It is only safe to call for one file at a time.
+type FileCopier interface {
+	Copy(ctx context.Context, file File) error
+	Close(ctx context.Context) error
+}
+
+// A FileCopyFactory is used to create FileCopiers when the CopyFilesSourceType is not
+// known until a later time. For example, asking to copy files from a machine results in
+// an RPC that is likely to be handled by a method that knows how to make FileCopiers, but
+// not what kind yet. This is a consequence of the current flow of operations and in the future
+// the factory may become obviated.
+type FileCopyFactory interface {
+	MakeFileCopier(ctx context.Context, sourceType CopyFilesSourceType) (FileCopier, error)
+}
+
+// A FileReadCopier is mostly a tee-like utility to read all files from somewhere and
+// pass them through to some bound-to, underlying FileCopier. Similar to the FileCopyFactory,
+// this abstraction exists so as to promote re-use across client/server interactions while
+// reducing some extra abstractions like a FileReader/FileIterator.
+type FileReadCopier interface {
+	ReadAll(ctx context.Context) error
+	Close(ctx context.Context) error
+}
+
+// CopyFilesSourceType indicates the types of files being transmitted.
+type CopyFilesSourceType int
+
+const (
+	// CopyFilesSourceTypeSingleFile is just one normal file being copied.
+	CopyFilesSourceTypeSingleFile CopyFilesSourceType = iota
+	// CopyFilesSourceTypeSingleDirectory is one top-level directory being
+	// copied but the transmission may contain many files. This disambiguates
+	// where to place the directory versus multiple files. Multiple files always
+	// go within the target destination, but a single directory may be renamed
+	// at the top depending on if the target destination exists or not. Either
+	// way, the receiving side needs to know what is being worked with. This
+	// is a consequence of mimicking SCP and is admittedly a bit involved as well
+	// as probably annoying to implement correctly.
+	CopyFilesSourceTypeSingleDirectory
+	// CopyFilesSourceTypeMultipleFiles indicates multiple files will be transmitted.
+	CopyFilesSourceTypeMultipleFiles
```

Comment:
[q] Is it worth supporting copying both directories and multiple files if doing both of those things adds a significant amount of complexity that will be difficult to maintain? Maybe I'm misunderstanding your comment above, and it's just a necessity of implementing this feature. I'm just nervous about netcode having to continue to support this.

Reply:
Although the comment make it sound not great, the only place where maintenance of this happens is figuring out how to treat the target, top-level directory in `localFileCopyFactory.MakeFileCopier`. It's fairly well documented. In terms of is it worth it, I think it comes down to a UX decision. I believe supporting mixing and matching of source files to a destination in the same way that scp does and it doing what you expect it to is pretty nice. The nerves make sense and FWIW, if it ends up needing a lot of maintenance is also complicated/cumbersome, it can be deprecated over time.

Reply:
>  believe supporting mixing and matching of source files to a destination in the same way that scp does and it doing what you expect it to is pretty nice.

Yep I buy it; and the longer I look at the code the less scary it seems haha. Seems good to me, just thought I'd bring other people in on my anxiety 😄 .